aea/mail/base.py

```diff
@@ -392,38 +373,47 @@ def __init__(
                 bool(protocol_id) or bool(protocol_specification_id),
                 "Message is bytes object, protocol_id or protocol_specification_id must be provided!",
             )
+            if protocol_id is not None and protocol_specification_id is not None:
+                protocol_specification_id_ = self._get_protocol_specification_id_by_the_protocol_id(
+                    protocol_id
+                )
+                enforce(
+                    protocol_specification_id == protocol_specification_id_,
+                    "Protocol specification id mismatch.",
+                )
+            elif protocol_id is not None:
+                protocol_specification_id = self._get_protocol_specification_id_by_the_protocol_id(
+                    protocol_id
+                )
         elif isinstance(message, Message):
-            # protocol_id provided as an argument in priority
-            # use Message.protocol_id only if no protocol_id provided
             if message.protocol_id is None:
                 raise ValueError(  # pragma: nocover
                     f"message class {type(message)} has no protocol_id specified!"
                 )
+            if protocol_id is not None:
+                enforce(
+                    protocol_id == message.protocol_id, "Inconsistent args provided."
+                )
             protocol_id = message.protocol_id
+            protocol_specification_id = self._get_protocol_specification_id_by_the_protocol_id(
+                protocol_id
+            )
         else:
             raise ValueError(
                 f"Message type: {type(message)} is not supported!"
             )  # pragma: nocover
 
-        if not protocol_id:
-            protocol_id = self._get_protocol_id_by_protocol_specification_id(
-                cast(PublicId, protocol_specification_id)
-            )
-        if not protocol_specification_id:
-            # if no protocol_spcification_id, try to resolve by protocol_id
-            protocol_specification_id = self._get_protocol_specification_id_by_the_protocol_id(
-                protocol_id
-            )
+        if protocol_specification_id is None and not is_framework_set:
+            protocol_specification_id = protocol_id
+        if protocol_id is None and not is_framework_set:
+            protocol_id = protocol_specification_id
```

Comment:
We need to unfortunately introduce the flag and do this here. Otherwise every single use of the Envelope needs the registry. The alternative is to use a new type of Envelope framework side.